sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw

```diff
@@ -0,0 +1,267 @@
+    pub fn trunc(self) -> Self {
+        Self {
+            // first move to the right (divide by the denominator)
+            // to get rid of fractional part, than move to the
+            // left (multiply by the denominator), to ensure 
+            // fixed-point structure
+            value: (self.value >> 32) << 32,
+        }
+    }
+}
+
+impl UFP64 {
+    /// Returns the largest integer less than or equal to `self`.
+    pub fn floor(self) -> Self {
+        return self.trunc();
+    }
+
+    /// Returns the fractional part of `self`.
+    pub fn fract(self) -> Self {
+        Self {
+            // first move to the left (multiply by the denominator)
+            // to get rid of integer part, than move to the
+            // right (divide by the denominator), to ensure 
+            // fixed-point structure
+            value: (self.value << 32) >> 32,
+        }
+    }
+}
+
+impl UFP64 {
+    /// Returns the smallest integer greater than or equal to `self`.
+    pub fn ceil(self) -> Self {
+        if self.fract().value != 0 {
+            let res = self.trunc() + ~UFP64::from_uint(1);
+            return res;
+        }
+        return self;
+    }
+}
+
+impl UFP64 {
+    /// Returns the nearest integer to `self`. Round half-way cases away from
+    pub fn round(self) -> Self {
+        let floor = self.floor();
+        let ceil = self.ceil();
+        let diff_self_floor = self - floor;
+        let diff_ceil_self = ceil - self;
+
+        // check if we are nearer to the floor or to the ceiling
+        if diff_self_floor < diff_ceil_self {
+            return floor;
+        } else {
+            return ceil;
+        }
+    }
+}
+
+impl Root for UFP64 {
+    /// Sqaure root for UFP64
+    fn sqrt(self) -> Self {
+        let nominator_root = self.value.sqrt();
+        // Need to multiple over 2 ^ 16, as the sqare root of the denominator 
+        // is also taken and we need to ensure that the denominator is constant
+        let nominator = nominator_root << 16;
+        Self {
+            value: nominator,
+        }
+    }
+}
+
+impl Exponentiate for UFP64 {
+    /// Power function. x ^ exponent
+    fn pow(self, exponent: Self) -> Self {
+        let demoninator_power = ~UFP64::denominator();
+        let exponent_int = exponent.value >> 32;
+        let nominator_pow = ~U128::from(0, self.value).pow(~U128::from(0, exponent_int));
+        // As we need to ensure the fixed point structure 
+        // which means that the denominator is always 2 ^ 32
+        // we need to delete the nominator by 2 ^ (32 * exponent - 1)
+        // - 1 is the formula is due to denominator need to stay 2 ^ 32
+        let nominator = nominator_pow >> demoninator_power * (exponent_int - 1);
+
+        if nominator.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        Self {
+            value: nominator.lower,
+        }
+    }
+}
+
+trait Exponent {
```

Comment:
I think this should be exported to the standard library's math library and then imported into this file.

Reply:
Unfortunately no, due to this known and AFAIK yet not fixed issue

Reply:
Fixed accordingly https://github.com/FuelLabs/sway-libs/pull/31/commits/035c90b7baadbacb7e189a91a5d932b6b115124f